storage/imx6ull-flashnor/flashnor.c

```diff
@@ -0,0 +1,295 @@
+	printf("\t-r <id> <fs>           - mount root filesystem\n");
+	printf("\t\tid:     device id starting at 0\n");
+	printf("\t\tfs:     filesystem name\n");
+	printf("\t-n <prefix> <id>       - use given prefix and starting id for device files registration\n");
+	printf("\t\tprefix: device prefix\n");
+	printf("\t\tid:     device starting id, e.g. -n flash 4 registers devices /dev/flash4, /dev/flash5, etc.\n");
+	printf("\t-h | --help            - print this help message\n");
+}
+
+
+static void flashnor_exit(int sig)
+{
+	exit(EXIT_SUCCESS);
+}
+
+
+int main(int argc, char *argv[])
+{
+	struct option longopts[] = {
+		{ "ecspi", required_argument, NULL, 'e' },
+		{ "qspi", required_argument, NULL, 'q' },
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+	char path[32], *prefix = FLASHNOR_PREFIX;
+	unsigned int port, id = FLASHNOR_ID;
+	int err, arg, c;
+	storage_t *dev, *parent;
+	oid_t oid;
+	pid_t pid;
+
+	/* Wait for console */
+	while (write(1, "", 0) < 0)
+		usleep(10000);
+
+	/* Wait for rootfs */
+	while (lookup("/", NULL, &oid) < 0)
+		usleep(10000);
+
+	/* Daemonize server */
+	signal(SIGUSR1, flashnor_exit);
+
+	if ((pid = fork()) < 0)
+		exit(EXIT_FAILURE);
+
+	if (pid > 0) {
+		sleep(10);
+		exit(EXIT_FAILURE);
+	}
+
+	signal(SIGUSR1, flashnor_exit);
+
+	if (setsid() < 0)
+		exit(EXIT_FAILURE);
+
+	if ((err = storage_init(flashnor_msgloop, 16)) < 0) {
+		fprintf(stderr, "imx6ull-flashnor: failed to initialize server, err: %d\n", err);
+		return err;
+	}
+	port = err;
+
+	if ((err = storage_registerfs("meterfs", meterfs_mount, meterfs_umount, meterfs_handler)) < 0) {
+		fprintf(stderr, "imx6ull-flashnor: failed to register filesystem, err: %d\n", err);
+		return err;
+	}
+
+	if (argc > 1) {
+		while ((c = getopt_long(argc, argv, "e:q:p:r:n:h", longopts, NULL)) != -1) {
+			switch (c) {
+				case 'e':
+					if ((dev = malloc(sizeof(storage_t))) == NULL) {
+						err = -ENOMEM;
+						fprintf(stderr, "imx6ull-flashnor: failed to allocate device, err: %d\n", err);
+						return err;
+					}
+
+					if ((err = flashnor_ecspiInit(strtoul(optarg, NULL, 0), dev)) < 0) {
+						fprintf(stderr, "imx6ull-flashnor: failed to initialize device, err: %d\n", err);
+						return err;
+					}
+
+					if ((err = storage_add(dev)) < 0) {
+						fprintf(stderr, "imx6ull-flashnor: failed to register device, err: %d\n", err);
+						return err;
+					}
+
+					oid.port = port;
+					oid.id = err;
+					if (snprintf(path, sizeof(path), "/dev/%s%u", prefix, id++) >= sizeof(path)) {
+						err = -ENAMETOOLONG;
+						fprintf(stderr, "imx6ull-flashnor: failed to build device file path, err: %d\n", err);
+						return err;
+					}
+
+					if ((err = create_dev(&oid, path)) < 0) {
+						fprintf(stderr, "imx6ull-flashnor: failed to create device file, err: %d\n", err);
+						return err;
+					}
+					break;
+
+				case 'q':
+					/* TODO: add QuadSPI NOR flash support */
+					return -ENOTSUP;
+
+				case 'p':
```

Comment:
- we would probably want to add also support for `ptable` in the future
- I don't know if we should allow user to set partition IDs via params, maybe just the order of params should define them?

Reply:
To be done in next driver iteration.